pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid        
+        InvalidBlock,
+        /// Initial block is invalid
+        InvalidInitialBlock,
+        /// General error during client operations
+        Other,
+    }
+
+    // NOTE for storage:
+    // - header storage should be a ring buffer (i.e. we keep n last headers, ordered by the insertion time)
+    // - we keep a pointer to the last finalized header (to avoid deserializing the whole buffer)
+    // - insertion moves the pointer and updates the buffer
+    // or?
+    // https://substrate.recipes/ringbuffer.html
+
+    /// Hash of the last imported header from the bridged chain
+    #[pallet::storage]
+    #[pallet::getter(fn get_last_imported_block_hash)]
+    pub type LastImportedBlockHash<T: Config> = StorageValue<_, TendermintBlockHash, ValueQuery>;
+
+    /// Imported hashes "ordered" by their insertion time
+    /// Client keeps HeadersToKeep number of these at any time    
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_hash)]
+    pub type ImportedHashes<T: Config> = StorageMap<_, Identity, u32, TendermintBlockHash>;
+
+    /// Current ring buffer position
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_hashes_pointer)]
+    pub(super) type ImportedHashesPointer<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    /// Bridged chain Headers which have been imported by the client
+    /// Client keeps HeadersToKeep number of these at any time
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_block)]
+    pub(super) type ImportedBlocks<T: Config> =
+        StorageMap<_, Identity, TendermintBlockHash, LightBlockStorage, OptionQuery>;
+
+    impl<T: Config> Pallet<T> {
+        pub fn get_last_imported_block() -> Option<LightBlockStorage> {
+            match ImportedHashesPointer::<T>::get().checked_sub(1) {
+                Some(ptr) => match ImportedHashes::<T>::get(ptr) {
+                    Some(key) => ImportedBlocks::<T>::get(key),
+                    None => None,
+                },
+                None => None,
+            }
+        }
+    }
+
+    /// If true, stop the world
+    #[pallet::storage]
+    #[pallet::getter(fn is_halted)]
+    pub type IsHalted<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn get_options)]
+    pub type LightClientOptions<T: Config> = StorageValue<_, LightClientOptionsStorage, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        // TODO : benchmark & adjust weights
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn initialize_client(
+            origin: OriginFor<T>,
+            options: LightClientOptionsStorage,
+            initial_block: LightBlockStorage,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
```

Comment:
This will have to be replaced with some other approach before releasing. We cannot require the `origin` to be a root/sudo since that would mean our relayer has to be given sudo keys. Which is not safe.

Let's make a ticket for it as well.

Reply:
but why does a relayer need a sudo key? Someone in the foundation creates the Client instance, and from that point on relayer can start submitting consecutive blocks (update call needs just a regular signature, can be called by any account).